src/Mod/Sketcher/Gui/Command.cpp

```diff
@@ -978,16 +977,290 @@ void CmdSketcherViewSection::activated(int iMsg)
 
 bool CmdSketcherViewSection::isActive()
 {
-    Gui::Document *doc = getActiveGuiDocument();
-    if (doc) {
-        // checks if a Sketch Viewprovider is in Edit and is in no special mode
-        SketcherGui::ViewProviderSketch* vp = dynamic_cast<SketcherGui::ViewProviderSketch*>(doc->getInEdit());
-        if (vp /*&& vp->getSketchMode() == ViewProviderSketch::STATUS_NONE*/)
-            return true;
+    return isEditModeToolActive(getActiveGuiDocument());
+}
+
+/* Grid tool ------------------------------ Grid Activate */
+DEF_STD_CMD_A(CmdSketcherGridActivate)
+
+bool toggleGrid(Gui::Document* doc) {
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");

```

Comment:
The grid toggling should only affect the edited sketch but shouldn't change the user settings.

Reply:
Yes about that I wanted to discuss if we should keep sketch specific grid settings. But from forum discussion it seems that we should indeed.

However this is giving me a problem, when the toolbar is initialized, sketchview isn't initialized yet (I think). So I cannot initialize the commands to the sketch settings. Which is why so far I made the modifications system wide.
See in : CmdSketcherCompGrid::createAction I left a comment there

Reply:
But this is a problem you still have independent of this. Imagine you create two sketches where the grid is disabled. Now you edit the first and enable the grid by calling this command. The grid of the first sketch and the command are in a consistent state.

Close the sketch and edit the second sketch. The grid is still disabled but the command indicates that it's enabled. So, here you have an inconsistency. The command should fetch its state from the sketch that the user is starting to edit.

Reply:
Yes sorry, what I meant is that I know this part is wrong and needs fixing. But before trying to I wanted to know if we keep sketch-specific grid settings. Which from the forum discussion the answer is yes.

So now this needs fixing to be sketch specific. My issue is that I don't know how the command can fetch its state from the sketch.

Look in CmdSketcherCompGrid::createAction. I tried to access the sketch but failed (see commented code). I'm not sure why I can't, but I'd guess that sketchView is not initialized at this point.

Reply:
There is the same problem for individual commands to setup the correct icon. Right now the icon is set with
`sPixmap = hGrp->GetBool("GridAuto") ? "Sketcher_GridAuto" : "Sketcher_GridAuto_Deactivated";`
Which doesn't take into account the sketch setting.

Reply:
> So now this needs fixing to be sketch specific. My issue is that I don't know how the command can fetch its state from the sketch.

It should be possible to implement it in the command's isActive() method. This function will be triggered when changing the selection.

Reply:
Great thanks. I'll do that then.

Reply:
Is that not a problem if this code is run is run everytime isActive is run?

I just pushed the corrections. Now the tool saves only the setting for the sketch. It does not change the preferences.

Reply:
> Is that not a problem if this code is run is run everytime isActive is run?

This should be fine. isActive() is triggered when the selection changes, the active window changes and for a few other things.